File: cit-api/pipeline/serializers/opportunity.py
from rest_framework import serializers
from pipeline.models.preferred_development import PreferredDevelopment
from pipeline.models.general import Municipality
from pipeline.models.indian_reserve_band_name import IndianReserveBandName
import datetime

from pipeline.models.opportunity import Opportunity, PostSecondaryDistance, CommunityDistance, MunicipalityDistance, IndianReserveBandDistance, LakeDistance, RiverDistance, RoadsAndHighwaysDistance, AirportDistance, RailwayDistance, PortAndTerminalDistance, CustomsPortOfEntryDistance, ResearchCentreDistance, FirstResponderDistance, HospitalDistance

class OpportunityPostSecondarySerializer(serializers.ModelSerializer):
    class Meta:
        model=PostSecondaryDistance
        fields=('location_id', 'location_distance')

class OpportunityCommunitySerializer(serializers.ModelSerializer):
    class Meta:
        model=CommunityDistance
        fields=('community_id', 'community_distance')

class OpportunityMunicipalitySerializer(serializers.ModelSerializer):
    class Meta:
        model=MunicipalityDistance
        fields=('municipality_id', 'municipality_distance')
class OpportunityIndianReserveBandSerializer(serializers.ModelSerializer):
    class Meta:
        model=IndianReserveBandDistance
        fields=('reserve_id', 'reserve_distance')

class OpportunityLakeSerializer(serializers.ModelSerializer):
    class Meta:
        model=LakeDistance
        fields=('lake_id', 'lake_distance')

class OpportunityRiverSerializer(serializers.ModelSerializer):
    class Meta:
        model=RiverDistance
        fields=('river_id', 'river_distance')

class OpportunityRoadsAndHighwaysSerializer(serializers.ModelSerializer):
    class Meta:
        model=RoadsAndHighwaysDistance
        fields=('highway_id', 'highway_distance')
class OpportunityAirportSerializer(serializers.ModelSerializer):
    class Meta:
        model=AirportDistance
        fields=('airport_id', 'airport_distance')

class OpportunityRailwaySerializer(serializers.ModelSerializer):
    class Meta:
        model=RailwayDistance
        fields=('railway_id', 'railway_distance')

class OpportunityPortAndTerminalSerializer(serializers.ModelSerializer):
    class Meta:
        model=PortAndTerminalDistance
        fields=('port_id', 'port_distance')

class OpportunityCustomsPortOfEntrySerializer(serializers.ModelSerializer):
    class Meta:
        model=CustomsPortOfEntryDistance
        fields=('customs_port_id', 'customs_port_distance')

class OpportunityResearchCentreSerializer(serializers.ModelSerializer):
    class Meta:
        model=ResearchCentreDistance
        fields=('research_centre_id', 'research_centre_distance')
 
class OpportunityFirstResponderSerializer(serializers.ModelSerializer):
    class Meta:
        model=FirstResponderDistance
        fields=('first_responder_id', 'first_responder_distance')

class OpportunityHospitalSerializer(serializers.ModelSerializer):
    class Meta:
        model=HospitalDistance
        fields=('hospital_id', 'hospital_distance')

class MunicipalitySerializer(serializers.ModelSerializer):
    class Meta:
        model=MunicipalityDistance
        fields=('municipality_distance', 'municipality_id')

class IndianReserveBandSerializer(serializers.ModelSerializer):
    class Meta:
        model=IndianReserveBandDistance
        fields=('reserve_distance', 'reserve_id')


class OpportunitySerializer(serializers.ModelSerializer):
    opportunity_preferred_development = serializers.PrimaryKeyRelatedField(queryset=PreferredDevelopment.objects.all(), many=True, required=False)
    nearest_first_nations = serializers.PrimaryKeyRelatedField(queryset=IndianReserveBandName.objects.all(), many=True, required=False)
    nearest_municipalities = serializers.PrimaryKeyRelatedField(queryset=Municipality.objects.all(), many=True, required=False)

    nearest_first_nations_object = IndianReserveBandSerializer(many=True, required=False)
    nearest_municipalities_object = MunicipalitySerializer(many=True, required=False)

    nearest_community = OpportunityCommunitySerializer(required=False)
    nearest_post_secondary = OpportunityPostSecondarySerializer(required=False)
    nearest_coast_guard_station = OpportunityFirstResponderSerializer(required=False)
    nearest_ambulance_station = OpportunityFirstResponderSerializer(required=False) 
    nearest_police_station = OpportunityFirstResponderSerializer(required=False)
    nearest_fire_station = OpportunityFirstResponderSerializer(required=False)
    nearest_health_center = OpportunityHospitalSerializer(required=False)
    nearest_research_centre = OpportunityResearchCentreSerializer(required=False)
    nearest_customs_port_of_entry = OpportunityCustomsPortOfEntrySerializer(required=False)
    nearest_port = OpportunityPortAndTerminalSerializer(required=False)
    nearest_railway = OpportunityRailwaySerializer(required=False)
    nearest_airport = OpportunityAirportSerializer(required=False)
    nearest_highway = OpportunityRoadsAndHighwaysSerializer(required=False)
    nearest_river = OpportunityRiverSerializer(required=False)
    nearest_lake = OpportunityLakeSerializer(required=False)

    class Meta:
        model = Opportunity
        fields = (
            "id",
            "deleted",
            "opportunity_address",
            "opportunity_name",
            "geo_position",
            "approval_status",
            "date_created",
            "date_updated",
            "date_published",
            "business_contact_name",
            "business_contact_email",
            "opportunity_description",
            "environmental_information",
            "opportunity_link",
            "community_link",
            "regional_district_id",
            "parcel_ownership",
            "parcel_size",
            "pid",
            "parcel_geometry",
            "elevation_at_location",
            "soil_name",
            "soil_texture",
            "soil_drainage",
            "opportunity_sale_price",
            "opportunity_rental_price",
            "opportunity_road_connected",
            "opportunity_water_connected",
            "opportunity_water_capacity",
            "opportunity_sewer_connected",
            "opportunity_sewer_capacity",
            "opportunity_natual_gas_connected",
            "opportunity_natual_gas_capacity",
            "opportunity_electrical_connected",
            "opportunity_electrical_capacity",
            "private_note",
            "public_note",
            "last_admin",
            "land_use_zoning",
            "ocp_zoning_code",
            "opportunity_property_status",
            "opportunity_preferred_development",
            "nearest_port",
            "nearest_airport",
            "nearest_research_centre",
            "nearest_customs_port_of_entry",
            "nearest_transmission_line",
            "nearest_coast_guard_station",
            "nearest_ambulance_station",
            "nearest_police_station",
            "nearest_fire_station",
            "nearest_health_center",
            "nearest_railway",
            "nearest_highway",
            "nearest_river",
            "nearest_lake",
            "nearest_first_nations",
            "nearest_municipalities",
            "nearest_post_secondary",
            "nearest_first_nations_object",
            "nearest_municipalities_object",
            "nearest_community"
        )
    
    def create(self, validated_data):
        nearest_community = None
        if validated_data.get('nearest_community'):
            community = validated_data.pop('nearest_community')
            if community.get('community_id') is not None:
                nearest_community = CommunityDistance.objects.create(**community)
        
        nearest_post_secondary = None
        if validated_data.get('nearest_post_secondary'):
            post_secondary = validated_data.pop('nearest_post_secondary')
            if post_secondary.get('location_id') is not None:
                nearest_post_secondary = PostSecondaryDistance.objects.create(**post_secondary)
        
        nearest_coast_guard_station = None
        if validated_data.get('nearest_coast_guard_station'):
            coast_guard_station = validated_data.pop('nearest_coast_guard_station')
            if coast_guard_station.get('first_responder_id') is not None:
                nearest_coast_guard_station = FirstResponderDistance.objects.create(**coast_guard_station)
        
        nearest_ambulance_station = None
        if validated_data.get('nearest_ambulance_station'):
            ambulance_station = validated_data.pop('nearest_ambulance_station')
            if ambulance_station.get('first_responder_id') is not None:
                nearest_ambulance_station = FirstResponderDistance.objects.create(**ambulance_station) 
        
        nearest_police_station = None
        if validated_data.get('nearest_police_station'):
            police_station = validated_data.pop('nearest_police_station')
            if police_station.get('first_responder_id') is not None:
                nearest_police_station = FirstResponderDistance.objects.create(**police_station)
        
        nearest_fire_station = None
        if validated_data.get('nearest_fire_station'):
            fire_station = validated_data.pop('nearest_fire_station')
            if fire_station.get('first_responder_id') is not None:
                nearest_fire_station = FirstResponderDistance.objects.create(**fire_station)
        
        nearest_health_center = None
        if validated_data.get('nearest_health_center'):
            health_center = validated_data.pop('nearest_health_center')
            if health_center.get('hospital_id') is not None:
                nearest_health_center = HospitalDistance.objects.create(**health_center)
        
        nearest_research_centre = None
        if validated_data.get('nearest_research_centre'):
            research_centre = validated_data.pop('nearest_research_centre')
            if research_centre.get('research_centre_id') is not None:
                nearest_research_centre = ResearchCentreDistance.objects.create(**research_centre)
        
        nearest_customs_port_of_entry = None
        if validated_data.get('nearest_customs_port_of_entry'):
            customs_port_of_entry = validated_data.pop('nearest_customs_port_of_entry')
            if customs_port_of_entry.get('customs_port_id') is not None:
                nearest_customs_port_of_entry = CustomsPortOfEntryDistance.objects.create(**customs_port_of_entry)
                validated_data['nearest_customs_port_of_entry'] = nearest_customs_port_of_entry
        
        nearest_port = None
        if validated_data.get('nearest_port'):
            port = validated_data.pop('nearest_port')
            if port.get('port_id') is not None:
                nearest_port = PortAndTerminalDistance.objects.create(**port)
                validated_data['nearest_port'] = nearest_port
       
        nearest_railway = None
        if validated_data.get('nearest_railway'):
            railway = validated_data.pop('nearest_railway')
            if railway.get('railway_id') is not None:
                nearest_railway = RailwayDistance.objects.create(**railway)
                validated_data['nearest_railway'] = nearest_railway

        nearest_airport = None
        if validated_data.get('nearest_airport'):
            airport = validated_data.pop('nearest_airport')
            if airport.get('airport_id') is not None:
                nearest_airport = AirportDistance.objects.create(**airport)
                validated_data['nearest_airport'] = nearest_airport

        nearest_highway = None
        if validated_data.get('nearest_highway'):
            highway = validated_data.pop('nearest_highway')
            if highway.get('highway_id') is not None:
                nearest_highway = RoadsAndHighwaysDistance.objects.create(**highway)
                validated_data['nearest_highway'] = nearest_highway
        
        nearest_river = None
        if validated_data.get('nearest_river'):
            river = validated_data.pop('nearest_river')
            if river.get('river_id') is not None:
                nearest_river = RiverDistance.objects.create(**river)
                validated_data['nearest_river'] = nearest_river
        
        nearest_lake = None
        if validated_data.get('nearest_lake'):
            lake = validated_data.pop('nearest_lake')
            if lake.get('lake_id') is not None:
                nearest_lake = LakeDistance.objects.create(**lake)
                validated_data['nearest_lake'] = nearest_lake

        preferred_developments = []
        if validated_data.get('opportunity_preferred_development'):
            preferred_developments = validated_data.pop('opportunity_preferred_development')

            
        # get reserves
        filtered_first_nations_distances = []
        if validated_data.get('nearest_first_nations_object'):
            first_nations = validated_data.pop('nearest_first_nations_object')
            # TODO: validation filtered_first_nations = IndianReserveBandName.objects.filter(name__in=first_nations)

            # create distances
            
            for nation in first_nations:
                # create and append 
                filtered_first_nations_distance = IndianReserveBandDistance.objects.create(**nation)
                filtered_first_nations_distances.append(filtered_first_nations_distance.pk)

        # get nearest municipalities
        filtered_municipality_distances = []
        if validated_data.get('nearest_municipalities_object'):
            municipalities = validated_data.pop('nearest_municipalities_object')
            # filtered_municipalities = Municipality.objects.filter(name__in=municipalities)
            
            
            for muni in municipalities:
                municipalities_distance = MunicipalityDistance.objects.create(**muni)
                filtered_municipality_distances.append(municipalities_distance.pk)

        # insert opportuntity with literal fields
        instance = Opportunity.objects.create(**validated_data)

        instance.opportunity_preferred_development.set(preferred_developments)

        if filtered_first_nations_distances:
            for dist in filtered_first_nations_distances:
                instance.nearest_first_nations.add(dist)
        
        if filtered_municipality_distances:
            for dist in filtered_municipality_distances:
                instance.nearest_municipalities.add(dist)

        instance.nearest_post_secondary = nearest_post_secondary
        instance.nearest_community = nearest_community
        instance.nearest_coast_guard_station = nearest_coast_guard_station
        instance.nearest_ambulance_station = nearest_ambulance_station
        instance.nearest_police_station = nearest_police_station
        instance.nearest_fire_station = nearest_fire_station
        instance.nearest_health_center = nearest_health_center
        instance.nearest_research_center = nearest_research_centre
        instance.nearest_customs_port_of_entry = nearest_customs_port_of_entry
        instance.nearest_port = nearest_port
        instance.nearest_railway = nearest_railway
        instance.nearest_airport = nearest_airport
        instance.nearest_highway = nearest_highway
        instance.nearest_river = nearest_river
        instance.nearest_lake = nearest_lake
        instance.save()

        return instance

File: cit3.0-web/src/store/models/opportunity.js
import _ from "lodash";
import { createOpportunityLink } from "../../helpers/helpers";

/**
 * Initial opportunity model
 */
export const OPPORTUNITY_MODEL = () => ({
  name: "",
  address: "",
  coords: [54.1722, -124.1207],
  approvalStatus: "",
  businessContactName: "",
  businessContactEmail: "",
  publicNote: "",
  privateNote: "",
  lastAdmin: "",
  user: 1,
  dateCreated: "",
  dateUpdated: "",
  datePublished: null,
  resourceIds: {
    Hospitals: "5ff82cf4-0448-4063-804a-7321f0f2b4c6",
    Schools: "5832eff2-3380-435e-911b-5ada41c1d30b",
    "Post Secondary Schools": "8e4e2a87-2d1d-4931-828e-6327b49f310e",
    Courts: "23aa0b75-2715-4ccb-9a36-9a608450dc2d",
    "Walk-In Clinics": "3ca6b086-c92b-4654-ae82-ff5723d00611",
    "Natural Resource Projects": "2b69cc4b-4076-4272-a5a0-1c731455e063",
    "Economic Projects": "b12cd4cc-b58b-4079-b630-a20b6df58e8d",
  },
  nearbyResources: {},
  community: {},
  regionalDistrict: {},
  municipality: {},
  municipalities: [],
  firstNationCommunities: [],
  services: {
    networkAvg: {
      title: "Site servicing",
      subtitle: "- Speed average",
      value: "",
      type: "paragraph",
    },
    networkAtRoad: {
      title: "Site servicing",
      subtitle: "- Speed at nearest road",
      value: "",
      type: "paragraph",
    },
    roadAccess: {
      title: "Site servicing",
      subtitle: "- Road Access",
      name: "Unknown",
      type: "text",
    },
    waterSupply: {
      title: "Site servicing",
      subtitle: "- Water",
      name: "Unknown",
      value: "",
      type: "capacity",
      suffix: "m³/hour",
    },
    naturalGas: {
      title: "Site servicing",
      subtitle: "- Natural Gas",
      name: "Unknown",
      value: "",
      type: "pressure",
      suffix: "MMBTU/hour",
    },
    sewer: {
      title: "Site servicing",
      subtitle: "- Sewer",
      name: "Unknown",
      value: "",
      type: "capacity",
      suffix: "m³/hour",
    },
    electrical: {
      title: "Site servicing",
      subtitle: "- Electrical",
      name: "Unknown",
      value: "",
      type: "capacity",
      suffix: "MW",
    },
    transmission: {
      title: "Power Transmission Lines",
      name: "No",
      value: 0,
      type: "distance",
      suffix: "km",
    },
    nearHealth: {
      title: "Health Care Facility within 100km",
      name: "No",
      value: 0,
      type: "distance",
      suffix: "km",
    },
    nearFire: {
      title: "First Responders within 100km",
      subtitle: "- Fire",
      name: "No",
      value: 0,
      type: "distance",
      suffix: "km",
    },
    nearAmbulance: {
      title: "First Responders within 100km",
      subtitle: "- Ambulance",
      name: "No",
      value: 0,
      type: "distance",
      suffix: "km",
    },
    nearPolice: {
      title: "First Responders within 100km",
      subtitle: "- Police",
      name: "No",
      value: 0,
      type: "distance",
      suffix: "km",
    },
    nearCoastGuard: {
      title: "First Responders within 100km",
      subtitle: "- Coast Guard",
      name: "No",
      value: 0,
      type: "distance",
      suffix: "km",
    },
    nearSecondarySchool: {
      title: "Post-Secondary Education Facility within 100km",
      name: "No",
      value: 0,
      type: "distance",
      suffix: "km",
    },
    nearResearchCentre: {
      title: "Research Centre within 100km",
      name: "No",
      value: 0,
      type: "distance",
      suffix: "km",
    },
  },
  transportation: {
    nearHighway: {
      title: "Nearest Highway",
      name: "",
      value: 0,
      type: "distance",
      suffix: "km",
    },
    nearAirport: {
      title: "Nearest Air Service",
      name: "",
      value: 0,
      type: "distance",
      suffix: "km",
    },
    nearRailway: {
      title: "Nearest Railway",
      name: "",
      value: 0,
      type: "distance",
      suffix: "km",
    },
    nearPort: {
      title: "Nearest Port",
      name: "",
      value: 0,
      type: "distance",
      suffix: "km",
    },
    nearCustomsPort: {
      title: "Nearest Canada Customs Port of Entry",
      name: "",
      value: 0,
      type: "distance",
      suffix: "km",
    },
  },
  physical: {
    nearElevation: {
      title: "Elevation at location",
      name: null,
      value: 0,
      type: "height",
      suffix: "m",
    },
    nearGround: {
      title: "Soil Name, Texture, Drainage",
      name: null,
      value: "",
      type: "text",
    },
    nearLake: {
      title: "Nearest lake",
      name: "",
      value: 0,
      type: "distance",
      suffix: "km",
    },
    nearRiver: {
      title: "Nearest river",
      name: "",
      value: 0,
      type: "distance",
      suffix: "km",
    },
  },
  siteInfo: {
    parcelOwnership: {
      title: "Ownership",
      name: "",
      type: "text",
    },
    parcelSize: {
      title: "Parcel size",
      value: null,
      type: "size",
      suffix: "acres",
    },
    PID: {
      title: "PID",
      value: null,
      type: "paragraph",
    },
    geometry: {
      title: "Polygon",
      type: "Polygon",
      coordinates: null,
      hidden: true,
    },
    siteId: {
      title: "Site ID",
      value: null,
      type: "text",
      hidden: true,
    },
  },
  userInfo: {
    saleOrLease: {
      title: "Sale or Lease",
      value: "",
      type: "select",
      rentalPrice: "",
      salePrice: "",
    },
    currentZone: { title: "Current Zoning", value: "", type: "select" },
    futureZone: { title: "Designation", value: "", type: "select" },
    preferredDevelopment: {
      title: "Preferred Development",
      value: [],
      type: "multi",
    },
    opportunityDescription: {
      title: "Opportunity Description",
      value: "",
      type: "paragraph",
    },
    environmentalInformation: {
      title: "Environmental Information",
      value: "",
      type: "paragraph",
    },
    communityLink: {
      title: "Visit Community website",
      value: "",
      type: "link",
    },
    opportunityLink: {
      title: "View listing for this Opportunity",
      value: "",
      type: "link",
    },
    editing: false,
  },
});

const STATUS_CODES = { Y: "Yes", N: "No", U: "Unknown" };

/**
 * Model used to map visual sections of the screens.
 */
export class Opportunity {
  constructor(state) {
    if (state) {
      this.state = _.mergeWith({}, state);
    } else {
      this.state = _.mergeWith({}, OPPORTUNITY_MODEL());
    }
  }

  set deleted(value) {
    this.state.deleted = value;
  }

  set editing(value) {
    this.state.editing = value;
  }

  set id(value) {
    this.state.id = value;
  }

  set approvalStatus(value) {
    this.state.approvalStatus = value;
  }

  set geoPosition(value) {
    const geo = value.match(/\((.*) (.*)\)/);
    this.state.coords = [parseFloat(geo[2]), parseFloat(geo[1])];
  }

  set opportunityAddress(value) {
    this.state.address = value;
  }

  set regionalDistrict(value) {
    this.state.regionalDistrict = value;
  }

  set municipality(value) {
    this.state.municipality = value;
  }

  set nearestMunicipalities(value) {
    if (value && value.features) {
      this.state.municipalities = value.features.map((feature) => ({
        name: feature.properties.name,
        link: feature.properties.link,
        distance: feature.properties.distance,
        population: feature.properties.population,
        pk: feature.properties.pk,
      }));
    } else {
      this.state.municipalities = value.map((feature) => ({
        name: feature.municipality_name,
        link: feature.municipality_link,
        distance: feature.municipality_distance,
        population: feature.municipality_population,
        pk: feature.municipality_id,
      }));
    }
  }

  set nearestFirstNations(value) {
    if (value && value.features) {
      this.state.firstNationCommunities = value.features.map((feature) => ({
        name: feature.properties.english_name,
        link: feature.properties.link,
        distance: feature.properties.distance,
        population: feature.properties.population,
        pk: feature.properties.pk,
      }));
    } else {
      this.state.firstNationCommunities = value.map((feature) => ({
        name: feature.reserve_name,
        link: feature.reserve_link,
        distance: feature.reserve_distance,
        population: feature.reserve_population,
        pk: feature.reserve_id,
      }));
    }
  }

  createLink() {
    this.state.link = createOpportunityLink(this.state.name, this.state.id);
  }

  set opportunityName(value) {
    this.state.name = value;
  }

  set dateCreated(value) {
    this.state.dateCreated = value;
  }

  set dateUpdated(value) {
    this.state.dateUpdated = value;
  }

  set publicNote(value) {
    this.state.publicNote = value;
  }

  set privateNote(value) {
    this.state.privateNote = value;
  }

  set lastAdmin(value) {
    this.state.lastAdmin = value;
  }

  set userId(value) {
    this.state.user = value;
  }

  // Business Contact
  set businessContactName(value) {
    this.state.businessContactName = value;
  }

  set businessContactEmail(value) {
    this.state.businessContactEmail = value;
  }

  // Transportation Info
  set nearestHighway(value) {
    if (value.features) {
      this.state.transportation.nearHighway.name =
        value.features && value.features[0].properties.name;
      this.state.transportation.nearHighway.pk =
        value.features && value.features[0].properties.pk;
    } else {
      this.state.transportation.nearHighway.name = value.name;
      this.state.transportation.nearHighway.value = value.highway_id;
    }
    this.state.transportation.nearHighway.value = value.highway_distance;
  }

  set nearestAirport(value) {
    if (value.features) {
      this.state.transportation.nearAirport.name =
        value.features[0].properties.name;
      this.state.transportation.nearAirport.pk =
        value.features[0].properties.pk;
    } else {
      this.state.transportation.nearAirport.name = value.name;
      this.state.transportation.nearAirport.pk = value.airport_id;
    }
    this.state.transportation.nearAirport.value = value.airport_distance;
  }

  set nearestRailway(value) {
    if (value.features) {
      this.state.transportation.nearRailway.name =
        value.features && value.features[0].properties.name;
      this.state.transportation.nearRailway.value = value.railway_distance;
      this.state.transportation.nearRailway.pk =
        value.features && value.features[0].properties.pk;
    } else {
      this.state.transportation.nearRailway.name = value.name;
      this.state.transportation.nearRailway.pk = value.railway_id;
    }
    this.state.transportation.nearRailway.value = value.railway_distance;
  }

  set nearestPort(value) {
    if (value.features) {
      this.state.transportation.nearPort.name =
        value.features && value.features[0].properties.name;
      this.state.transportation.nearPort.pk =
        value.features && value.features[0].properties.pk;
    } else {
      this.state.transportation.nearPort.name = value.name;
      this.state.transportation.nearPort.pk = value.railway_id;
    }
    this.state.transportation.nearPort.value = value.port_distance;
  }

  set nearestCustomsPortOfEntry(value) {
    if (value.features) {
      this.state.transportation.nearCustomsPort.name =
        value.features && value.features[0].properties.name;
      this.state.transportation.nearCustomsPort.pk =
        value.features && value.features[0].properties.pk;
    } else {
      this.state.transportation.nearCustomsPort.name = value.name;
      this.state.transportation.nearCustomsPort.pk = value.railway_id;
    }
    this.state.transportation.nearCustomsPort.value =
      value.customs_port_distance;
  }

  // Physical Info
  set nearElevation(value) {
    this.state.physical.nearElevation.value = parseFloat(value);
  }

  set elevationAtLocation(value) {
    this.state.physical.nearElevation.value = parseFloat(value);
  }

  set nearGround(value) {
    this.state.physical.nearGround.name = value;
  }

  set nearestLake(value) {
    if (value) {
      this.state.physical.nearLake.name = "Yes";
    }
    if (value.features) {
      this.state.physical.nearLake.pk =
        value.features && value.features[0].properties.pk;
    } else {
      this.state.physical.nearLake.pk = value.features && value.lake_id;
    }
    this.state.physical.nearLake.value = value.lake_distance;
  }

  set nearestRiver(value) {
    if (value) {
      this.state.physical.nearRiver.name = "Yes";
    }
    if (value.features) {
      this.state.physical.nearRiver.pk =
        value.features && value.features[0].properties.pk;
    } else {
      this.state.physical.nearRiver.pk = value.river_id;
    }
    this.state.physical.nearRiver.value = value.river_distance;
  }

  // Services
  set opportunityRoadConnected(name) {
    this.state.services.roadAccess.name = STATUS_CODES[name];
  }

  set opportunityWaterConnected(name) {
    this.state.services.waterSupply.name = STATUS_CODES[name];
  }

  set opportunityWaterCapacity(value) {
    this.state.services.waterSupply.value = value;
  }

  set opportunitySewerConnected(name) {
    this.state.services.sewer.name = STATUS_CODES[name];
  }

  set opportunitySewerCapacity(value) {
    this.state.services.sewer.value = value;
  }

  set opportunityNaturalGasConnected(name) {
    this.state.services.naturalGas.name = STATUS_CODES[name];
  }

  set opportunityNaturalGasCapacity(value) {
    this.state.services.naturalGas.value = value;
  }

  set opportunityElectricalConnected(name) {
    this.state.services.electrical.name = STATUS_CODES[name];
  }

  set opportunityElectricalCapacity(value) {
    this.state.services.electrical.value = value;
  }

  set nearestTransmissionLine(value) {
    if (value) {
      this.state.services.transmission.name = "Yes";
    }
    this.state.services.transmission.value = value;
  }

  set networkAvg(value) {
    this.state.services.networkAvg.value = value;
  }

  set networkAtRoad(value) {
    this.state.services.networkAtRoad.value = value;
  }

  set nearestTransmission(value) {
    if (value) {
      this.state.services.transmission.name = "Yes";
    }
    this.state.services.transmission.value = value.distance;
  }

  set nearestResearchCentre(value) {
    if (value) {
      this.state.services.nearResearchCentre.name = "Yes";
    }
    if (value.features) {
      this.state.services.nearResearchCentre.value =
        value.research_centre_distance;
      this.state.services.nearResearchCentre.pk =
        value.features && value.features[0].properties.pk;
    } else {
      this.state.services.nearResearchCentre.value = parseFloat(
        value.research_centre_distance
      );
      this.state.services.nearResearchCentre.pk = value.research_centre_id;
    }
  }

  set nearestHealthCenter(value) {
    if (value) {
      this.state.services.nearHealth.name = "Yes";
    }
    if (value.features) {
      this.state.services.nearHealth.value = value.distance;
      this.state.services.nearHealth.pk =
        value.features && value.features[0].properties.pk;
    } else {
      this.state.services.nearHealth.value = value.hospital_distance;
      this.state.services.nearHealth.pk = value.hospital_id;
    }
  }

  set nearestFireStation(value) {
    if (value) {
      this.state.services.nearFire.name = "Yes";
    }
    if (value.features) {
      this.state.services.nearFire.value = value.distance;
      this.state.services.nearFire.pk =
        value.features && value.features[0].properties.pk;
    } else {
      this.state.services.nearFire.value = value.first_responder_distance;
      this.state.services.nearFire.pk = value.first_responder_id;
    }
  }

  set nearestAmbulanceStation(value) {
    if (value) {
      this.state.services.nearAmbulance.name = "Yes";
    }
    if (value.features) {
      this.state.services.nearAmbulance.value = value.distance;
      this.state.services.nearAmbulance.pk =
        value.features && value.features[0].properties.pk;
    } else {
      this.state.services.nearAmbulance.value = value.first_responder_distance;
      this.state.services.nearAmbulance.pk = value.first_responder_id;
    }
  }

  set nearestPoliceStation(value) {
    if (value) {
      this.state.services.nearPolice.name = "Yes";
    }
    if (value.features) {
      this.state.services.nearPolice.value = value.distance;
      this.state.services.nearPolice.pk =
        value.features && value.features[0].properties.pk;
    } else {
      this.state.services.nearPolice.value = value.first_responder_distance;
      this.state.services.nearPolice.pk = value.first_responder_id;
    }
  }

  set nearestCoastGuardStation(value) {
    if (value) {
      this.state.services.nearCoastGuard.name = "Yes";
    }
    if (value.features) {
      this.state.services.nearCoastGuard.value = value.distance;
      this.state.services.nearCoastGuard.pk =
        value.features && value.features[0].properties.pk;
    } else {
      this.state.services.nearCoastGuard.value = value.first_responder_distance;
      this.state.services.nearCoastGuard.pk = value.first_responder_id;
    }
  }

  set nearestPostSecondary(value) {
    if (value) {
      this.state.services.nearSecondarySchool.name = "Yes";
    }
    if (value.features) {
      this.state.services.nearSecondarySchool.value =
        value.post_secondary_distance;
      this.state.services.nearSecondarySchool.pk =
        value.features && value.features[0].properties.pk;
    } else {
      this.state.services.nearSecondarySchool.value = value.location_distance;
      this.state.services.nearSecondarySchool.pk = value.location_id;
    }
  }

  // Site Info
  set parcelOwnership(value) {
    this.state.siteInfo.parcelOwnership.name = value;
  }

  get parcelOwnership() {
    return this.state.siteInfo.parcelOwnership.name;
  }

  set parcelSize(value) {
    this.state.siteInfo.parcelSize.value = parseFloat(value);
  }

  set PID(value) {
    this.state.siteInfo.PID.value = [value];
  }

  set parcelGeometry(value) {
    this.state.siteInfo.geometry.coordinates = value;
  }

  set geometry(value) {
    this.state.siteInfo.geometry.polygon = value;
  }

  // User Info
  set saleOrLease(value) {
    this.state.userInfo.saleOrLease.value = value;
  }

  set opportunitySalePrice(value) {
    this.state.userInfo.saleOrLease.salePrice = value;
  }

  set opportunityRentalPrice(value) {
    this.state.userInfo.saleOrLease.rentalPrice = value;
  }

  set rentalPrice(value) {
    this.state.userInfo.saleOrLease.rentalPrice = value;
  }

  set salePrice(value) {
    this.state.userInfo.saleOrLease.salePrice = value;
  }

  set ocpZoningCode(value) {
    this.state.userInfo.futureZone.value = value;
  }

  set landUseZoning(value) {
    this.state.userInfo.currentZone.value = value;
  }

  set opportunityPreferredDevelopmentV2(value) {
    this.state.userInfo.preferredDevelopment.value = value
      .split(";")
      .map((label) => ({ label }));
  }

  set opportunityPropertyStatus(value) {
    this.state.userInfo.saleOrLease.value = value;
  }

  set opportunityDescription(value) {
    this.state.userInfo.opportunityDescription.value = value;
  }

  set environmentalInformation(value) {
    this.state.userInfo.environmentalInformation.value = value;
  }

  set communityLink(value) {
    this.state.userInfo.communityLink.value = value;
  }

  set community(value) {
    this.state.community.id = value.features && value.features[0].properties.pk;
    this.state.community.community_distance = value.community_distance;
  }

  set opportunityLink(value) {
    this.state.userInfo.opportunityLink.value = value;
  }

  asState() {
    return this.state;
  }
}
